File: week09/19/P2.py
class Solution(object):
    def isBipartite(self, graph):
        visited = [0] * len(graph)
        
        def dfs_util(v, color):
            bipartite = True
            visited[v] = color
            for neighbor in graph[v]:
                if visited[neighbor] == color:
                    return False
                elif visited[neighbor] == 0:
                    bipartite = dfs_util(neighbor, -color)
            return bipartite

        for i in range(len(graph)):
            if (visited[i] == 0 and not dfs_util(i, 1)):
                return False
        return True

if __name__ == "__main__":
    graph = [[1,3],[0,2],[1,3],[0,2]]
    result = Solution()
    print(result.isBipartite(graph))
    graph = [[],[2,4,6],[1,4,8,9],[7,8],[1,2,8,9],[6,9],[1,5,7,8,9],[3,6,9],[2,3,4,6,9],[2,4,5,6,7,8]]
    print(result.isBipartite(graph))

File: week09/19/P2.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

bool dfs_util(int** graph, int* visited, int v, int color, int* graphColSize) {
    visited[v] = color;
    for (int k=0; k<graphColSize[v]; k++) {
        if (visited[graph[v][k]] == color) return false;
        else if (visited[graph[v][k]] == 0) {
            if (!dfs_util(graph, visited, graph[v][k], -color, graphColSize)) return false;
        }
    }
    return true;
}

bool isBipartite(int** graph, int graphSize, int* graphColSize) {
    int visited[graphSize];
    for (int i=0; i<graphSize; i++) visited[i] = 0;
    for (int j=0; j<graphSize; j++) {
        if (visited[j] == 0 && !(dfs_util(graph, visited, j, 1, graphColSize))) return false;
    }
    return true;
}

int main() {
    int arr1[] = {1,3};
    int arr2[] = {0,2};
    int arr3[] = {1,3};
    int arr4[] = {0,2};
    int *graph[] = {arr1, arr2, arr3, arr4};
    int graphSize = sizeof(graph)/sizeof(graph[0]);
    int graphColSize[graphSize];
    for (int i=0; i<graphSize; i++) graphColSize[i] = sizeof(*graph[i])/sizeof(int);
    printf("%d\n", isBipartite(graph, graphSize, graphColSize));
    
    return 0;
}